hedera-node/hapi-utils/src/main/java/com/hedera/node/app/hapi/utils/throttles/DiscreteLeakyBucket.java

```diff
@@ -1,78 +1,121 @@
+ *      the `nominalCapacityFree()` method to decide whether some action is allowed or not
+ * */
 public class DiscreteLeakyBucket {
     private long used;
-    private final long capacity;
+    private final long nominalCapacity;
+    private final long brimfulCapacity;
+
+    /**
+     * Creates a new bucket of both nominal and brimful capacities equal to {@param capacity}.
+     */
+    public static DiscreteLeakyBucket ofFixedCapacity(long capacity) {
+        return new DiscreteLeakyBucket(capacity, capacity);
+    }
+
+    /**
+     * Creates a new bucket of the given nominal and brimful capacities.
+     */
+    public static DiscreteLeakyBucket ofNominalAndBrimfulCapacity(long nominalCapacity, long brimfulCapacity) {
+        return new DiscreteLeakyBucket(nominalCapacity, brimfulCapacity);
+    }
+
+    private DiscreteLeakyBucket(long nominalCapacity, long brimfulCapacity) {
+        assertArgument(nominalCapacity >= 0, "nominalCapacity must be >= 0");
+        assertArgument(brimfulCapacity >= 0, "brimfulCapacity must be >= 0");
 
-    DiscreteLeakyBucket(long capacity) {
         used = 0L;
-        assertValidState(used, capacity);
-        this.capacity = capacity;
+        this.nominalCapacity = nominalCapacity;
+        this.brimfulCapacity = brimfulCapacity;
     }
 
-    /* Used only for test setup */
-    DiscreteLeakyBucket(long used, long capacity) {
-        assertValidState(used, capacity);
-        this.used = used;
-        this.capacity = capacity;
+    /**
+     * @return an absolute value of available nominal capacity. Returns 0 when overfilled.
+     */
+    public long nominalCapacityFree() {
+        return Math.max(0, nominalCapacity - used);
     }
 
-    long capacityFree() {
-        return capacity - used;
+    /**
+     * @return an absolute value of available brimful capacity. Returns 0 when full. Brimful capacity can't overfill.
+     */
+    public long brimfulCapacityFree() {
+        return Math.max(0, brimfulCapacity - used);
     }
 
-    long capacityUsed() {
+    /**
+     * @return an absolute value of used capacity.
+     */
+    public long capacityUsed() {
         return used;
     }
 
-    public long totalCapacity() {
-        return capacity;
+    /**
+     * @return the nominal capacity of this bucket
+     */
+    public long nominalCapacity() {
+        return nominalCapacity;
+    }
+
+    /**
+     * @return the brimful capacity of this bucket
+     */
+    public long brimfulCapacity() {
+        return brimfulCapacity;
     }
 
-    void useCapacity(long units) {
+    /**
+     * Uses a given amount of capacity. Throws an IllegalArgumentException if brimful capacity was to be exceeded.
+     */
+    public void useCapacity(long units) {
+        assertArgument(units >= 0, "units to leak must be >= 0");
+
         long newUsed = used + units;
-        assertValidUsage(units, newUsed);
+        assertArgument(
```

Comment:
This is a very hot code path, so I would prefer not to introduce a varargs method call into it (i.e., this allocates an almost-always-unused array for every invocation).